wdl-lint/src/v1/snake_case.rs

```diff
@@ -0,0 +1,229 @@
+//! A lint rule for ensuring tasks, workflows, and variables are named using
+//! snake_case.
+
+use std::fmt;
+
+use convert_case::Boundary;
+use convert_case::Case;
+use convert_case::Converter;
+use wdl_ast::experimental::v1::BoundDecl;
+use wdl_ast::experimental::v1::InputSection;
+use wdl_ast::experimental::v1::OutputSection;
+use wdl_ast::experimental::v1::StructDefinition;
+use wdl_ast::experimental::v1::TaskDefinition;
+use wdl_ast::experimental::v1::UnboundDecl;
+use wdl_ast::experimental::v1::Visitor;
+use wdl_ast::experimental::v1::WorkflowDefinition;
+use wdl_ast::experimental::AstToken;
+use wdl_ast::experimental::Diagnostic;
+use wdl_ast::experimental::Diagnostics;
+use wdl_ast::experimental::Span;
+use wdl_ast::experimental::VisitReason;
+
+use super::Rule;
+use crate::Tag;
+use crate::TagSet;
+
+/// Represents context of an warning.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum Context {
+    /// The warning occurred in a task.
+    Task,
+    /// The warning occurred in a workflow.
+    Workflow,
+    /// The warning occurred in a struct.
+    Struct,
+    /// The warning occurred in an input section.
+    Input,
+    /// The warning occurred in an output section.
+    Output,
+    /// The warning occurred in a private declaration.
+    PrivateDecl,
+}
+
+impl fmt::Display for Context {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Task => write!(f, "task"),
+            Self::Workflow => write!(f, "workflow"),
+            Self::Struct => write!(f, "struct member"),
+            Self::Input => write!(f, "input"),
+            Self::Output => write!(f, "output"),
+            Self::PrivateDecl => write!(f, "private declaration"),
+        }
+    }
+}
+
+/// The identifier for the snake_case rule.
+const ID: &str = "SnakeCase";
+
+/// Creates a "snake case" diagnostic.
+fn snake_case(context: Context, name: &str, properly_cased_name: &str, span: Span) -> Diagnostic {
+    Diagnostic::warning(format!("{context} identifier `{name}` is not snake_case"))
```

Comment:
Nit: I think from the user's perspective these are "names" (although from the grammar they're identifiers):

```suggestion
    Diagnostic::warning(format!("{context} name `{name}` is not snake_case"))
```